services/upgrader/upgrader.go

```diff
@@ -0,0 +1,179 @@
+package upgrader
+
+import (
+	"archive/tar"
+	"compress/gzip"
+	"errors"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"path/filepath"
+	"runtime"
+	"sync/atomic"
+
+	"github.com/mattermost/mattermost-server/v5/model"
+)
+
+var upgradePercentage int64
+var upgradeError error
+var upgrading int32
+
+type writeCounter struct {
+	total  int64
+	readed int64
+}
+
+func (wc *writeCounter) Write(p []byte) (int, error) {
+	n := len(p)
+	wc.readed += int64(n)
+	percentage := (wc.readed * 100) / wc.total
+	if percentage == 0 {
+		upgradePercentage = 1
+	} else if percentage == 100 {
+		upgradePercentage = 99
+	} else {
+		upgradePercentage = percentage
+	}
+	return n, nil
+}
+
+func getUpgradeDirectory() (string, error) {
+	exePath, err := os.Executable()
+	if err != nil {
+		return "", err
+	}
+	return filepath.Dir(exePath), nil
+}
+
+func getCurrentVersionTgzUrl() string {
+	return "https://releases.mattermost.com/" + model.CurrentVersion + "/mattermost-" + model.CurrentVersion + "-linux-amd64.tar.gz"
+}
+
+func canIUpgrade() error {
+	if runtime.GOARCH != "amd64" {
+		return errors.New("Upgrades only supported for amd64 architectures.")
+	}
+	if runtime.GOOS != "linux" {
+		return errors.New("Upgrades only supported for linux operating systems.")
+	}
+	return nil
+}
+
+func canIUpgradeToE0() error {
+	if err := canIUpgrade(); err != nil {
+		return err
+	}
+	if model.BuildEnterpriseReady == "true" {
+		return errors.New("You can't upgrade your code to enterprise because you are already in enterprise code")
+	}
+	return nil
+}
+
+func UpgradeToE0() error {
+	if !atomic.CompareAndSwapInt32(&upgrading, 0, 1) {
+		return errors.New("One upgrade is already running.")
+	}
+	defer atomic.CompareAndSwapInt32(&upgrading, 1, 0)
+
+	upgradePercentage = 1
+	upgradeError = nil
+
+	if err := canIUpgradeToE0(); err != nil {
```

Comment:
would it make sense to call this at the top of this method?

Reply:
yes, probably can be cleaner.

Reply:
Done